Create  a class ‘Car261_264’ which takes a url/path as an input to create an object. Assert if the columns in the data frame are same as [  
* id_1
* id_2
* route
* moto
* car
* rv
* bus
* truck

] and in the same order. This class should have five methods (names given below) which performs the said functionality.


In [139]:
import pandas as pd,numpy as np

class Car261_264:
    def __init__(self, url):
        self.df = pd.read_csv(r"C:\Users\Sam\Downloads\dataset-1.csv")
        self.assert_columns()

    def assert_columns(self):
        expected_columns = ["id_1", "id_2", "route", "moto", "car", "rv", "bus", "truck"]
        assert list(self.df.columns) == expected_columns, "Columns do not match."

    def generate_car_matrix(self):
        
        columns = self.df['id_2'].unique()
        index = self.df['id_1'].unique()
        car_matrix = pd.DataFrame(index=index, columns=columns)
        for _, row in self.df.iterrows():
            car_matrix.at[row['id_1'], row['id_2']] = row['car']
        return car_matrix
    
    def get_type_count(self):
        self.assert_columns() 
        self.df['car_type'] = pd.cut(self.df['car'], bins=[-float('inf'), 15, 25, float('inf')],
                                     labels=['low', 'medium', 'high'], include_lowest=True, right=False)
        type_count = self.df['car_type'].value_counts().to_dict()
        type_count = dict(sorted(type_count.items()))
        return type_count

    def get_bus_indexes(self):
        self.assert_columns()
        mean_bus = 2 * self.df['bus'].mean()
        bus_indexes = self.df[self.df['bus'] > mean_bus].index.tolist()
        return sorted(bus_indexes)

 
    def filter_routes(self):
        self.assert_columns()
        filtered_routes = self.df.groupby('route')['truck'].mean()
        filtered_routes = filtered_routes[filtered_routes > 7].index.tolist()
        return sorted(filtered_routes)


    def multiply_matrix(self, car_matrix):
        modified_matrix = car_matrix.applymap(lambda x: x * 0.75 if x > 20 else x * 1.25).round(1)
        return modified_matrix


# Question 1: Car Matrix Generation
* Under the method named generate_car_matrix write a logic that takes the dataset-1.csv as a DataFrame. Return a new DataFrame that follows the following rules:
* values from id_2 as columns
* values from id_1 as index
* dataframe should have values from car column
* diagonal values should be 0.


In [140]:
car_instance = Car261_264(r"C:\Users\Sam\Downloads\dataset-1.csv")

In [141]:
result_matrix = car_instance.generate_car_matrix().fillna(0)
result_matrix

827    821    804    822    826    807    809    801    824    823  \
829   4.14  13.26  28.92  11.81   5.81  19.62  15.72  36.32  21.40  10.00   
827   0.00   9.43  25.13   8.00   2.05  15.83  11.93  32.53   4.92   6.15   
821   9.43   0.00  16.12   1.80   8.01   6.82   2.92  23.52   5.06   3.67   
804  25.13  16.12   0.00  17.27  23.47   9.50  13.70   7.70  20.52  19.13   
822   8.00   1.80  17.27   0.00   6.55   7.97   4.07  24.67   3.60   2.21   
826   2.05   8.01  23.47   6.55   0.00  14.17  10.27  30.87   3.50   4.74   
807  15.83   6.82   9.50   7.97  14.17   0.00   4.30  16.90  11.22   9.83   
809  11.93   2.92  13.70   4.07  10.27   4.30   0.00  21.00   7.32   5.93   
801  32.53  23.52   7.70  24.67  30.87  16.90  21.00   0.00  27.92  26.53   
824   4.92   5.06  20.52   3.60   3.50  11.22   7.32  27.92   0.00   1.79   
823   6.15   3.67  19.13   2.21   4.74   9.83   5.93  26.53   1.79   0.00   
831   7.04  16.15  31.84  14.68   8.69  22.54  18.64  39.24  11.64  12.87   
808  13.13   4.12  12.30   5.27  11.47   2.90   1.70  19.60   8.52   7.13   
802  29.93  20.92   5.20  22.07  28.27  14.30  18.50   2.80  25.32  23.93   
805  21.03  12.02   4.40  13.17  19.37   5.40   9.60  11.70  16.42  15.03   
825   3.65   6.22  21.68   4.76   2.20  12.38   8.48  29.08   1.71   2.94   
806  19.43  10.42   6.10  11.57  17.77   3.80   8.00  13.40  14.82  13.43   
830   6.06  15.17  30.87  13.74   7.71  21.57  17.67  38.27  10.66  11.89   
803  26.73  17.72   2.00  18.87  25.07  11.10  15.30   6.00  22.12  20.73   

       831    808    802    805    825    806    830    803    829  
829   3.36  16.92  33.72  24.82   7.46  23.22   2.38  30.52   0.00  
827   7.04  13.13  29.93  21.03   3.65  19.43   6.06  26.73   4.14  
821  16.15   4.12  20.92  12.02   6.22  10.42  15.17  17.72  13.26  
804  31.84  12.30   5.20   4.40  21.68   6.10  30.87   2.00  28.92  
822  14.68   5.27  22.07  13.17   4.76  11.57  13.74  18.87  11.81  
826   8.69  11.47  28.27  19.37   2.20  17.77   7.71  25.07   5.81  
807  22.54   2.90  14.30   5.40  12.38   3.80  21.57  11.10  19.62  
809  18.64   1.70  18.50   9.60   8.48   8.00  17.67  15.30  15.72  
801  39.24  19.60   2.80  11.70  29.08  13.40  38.27   6.00  36.32  
824  11.64   8.52  25.32  16.42   1.71   7.80  10.66  22.12   8.77  
823  12.87   7.13  23.93  15.03   2.94  13.43  11.89  20.73  10.00  
831   0.00  19.84  36.64  27.74  10.33  26.14   1.39  33.44   3.36  
808  19.84   0.00  17.10   8.20   9.68   6.60  18.87  13.90  16.92  
802  36.64  17.10   0.00   9.20  26.48  10.90  35.67   3.40  33.72  
805  27.74   8.20   9.20   0.00  17.58   2.00  26.77   6.00  24.82  
825  10.33   9.68  26.48  17.58   0.00  15.98   9.35  23.28   7.46  
806  26.14   6.60  10.90   2.00  15.98   0.00  25.17   7.70  23.22  
830   1.39  18.87  35.67  26.77   0.00  25.17   0.00  32.47   2.38  
803  33.44  13.90   3.40   6.00  23.28   7.70  32.47   0.00  30.52

# Question 2: Car Type Count Calculation
Create a Python method named get_type_count that takes the dataset-1.csv as a DataFrame. Add a new categorical column car_type based on values of the column car:

* low for values less than or equal to 15,
* medium for values greater than 15 and less than or equal to 25,
* high for values greater than 25.
Calculate the count of occurrences for each car_type category and return the result as a dictionary. Sort the dictionary alphabetically based on keys.


In [126]:
car_instance = Car261_264(r"C:\Users\Sam\Downloads\dataset-1.csv")

In [127]:
type_count_result = car_instance.get_type_count()
type_count_result

{'high': 56, 'low': 196, 'medium': 89}

* The car values which are less than or equal to 15 have been occured 196 times
* The car values which are greater than 15 and less than or equal to 25 have been occured 89 times
* The car values which are greater than 25 have been occured 56 times




# Question 3: Bus Count Index Retrieval
* Create a Python method named get_bus_indexes that takes the dataset-1.csv as a DataFrame. The function should identify and return the indices as a list (sorted in ascending order) where the bus values are greater than twice the mean value of the bus column in the DataFrame.


In [128]:
car_instance = Car261_264(r"C:\Users\Sam\Downloads\dataset-1.csv")

In [129]:
bus_indexes = car_instance.get_bus_indexes()
print(bus_indexes)

[2, 7, 12, 17, 25, 30, 54, 64, 70, 97, 144, 145, 149, 154, 160, 201, 206, 210, 215, 234, 235, 245, 250, 309, 314, 319, 322, 323, 334, 340]


# Question 4: Route Filtering
* Create a python method filter_routes that takes the dataset-1.csv as a DataFrame. The function should return the sorted list of values of column route for which the average of values of truck column is greater than 7.

In [130]:
car_instance = Car261_264(r"C:\Users\Sam\Downloads\dataset-1.csv")

In [131]:
filtered_routes = car_instance.filter_routes()
print(filtered_routes)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


# Question 5: Matrix Value Modification
* Create a Python method named multiply_matrix that takes the resulting DataFrame from Question 1, as input and modifies each value according to the following logic:

* If a value in the DataFrame is greater than 20, multiply those values by 0.75,
* If a value is 20 or less, multiply those values by 1.25.

* The function should return the modified DataFrame which has values rounded to 1 decimal place.


In [137]:
car_instance = Car261_264(r"C:\Users\Sam\Downloads\dataset-1.csv")

In [149]:

# Test generate_car_matrix method
car_matrix = car_instance.generate_car_matrix().fillna(0)
print("Car Matrix:")
car_matrix

Car Matrix:


827    821    804    822    826    807    809    801    824    823  \
829   4.14  13.26  28.92  11.81   5.81  19.62  15.72  36.32  21.40  10.00   
827   0.00   9.43  25.13   8.00   2.05  15.83  11.93  32.53   4.92   6.15   
821   9.43   0.00  16.12   1.80   8.01   6.82   2.92  23.52   5.06   3.67   
804  25.13  16.12   0.00  17.27  23.47   9.50  13.70   7.70  20.52  19.13   
822   8.00   1.80  17.27   0.00   6.55   7.97   4.07  24.67   3.60   2.21   
826   2.05   8.01  23.47   6.55   0.00  14.17  10.27  30.87   3.50   4.74   
807  15.83   6.82   9.50   7.97  14.17   0.00   4.30  16.90  11.22   9.83   
809  11.93   2.92  13.70   4.07  10.27   4.30   0.00  21.00   7.32   5.93   
801  32.53  23.52   7.70  24.67  30.87  16.90  21.00   0.00  27.92  26.53   
824   4.92   5.06  20.52   3.60   3.50  11.22   7.32  27.92   0.00   1.79   
823   6.15   3.67  19.13   2.21   4.74   9.83   5.93  26.53   1.79   0.00   
831   7.04  16.15  31.84  14.68   8.69  22.54  18.64  39.24  11.64  12.87   
808  13.13   4.12  12.30   5.27  11.47   2.90   1.70  19.60   8.52   7.13   
802  29.93  20.92   5.20  22.07  28.27  14.30  18.50   2.80  25.32  23.93   
805  21.03  12.02   4.40  13.17  19.37   5.40   9.60  11.70  16.42  15.03   
825   3.65   6.22  21.68   4.76   2.20  12.38   8.48  29.08   1.71   2.94   
806  19.43  10.42   6.10  11.57  17.77   3.80   8.00  13.40  14.82  13.43   
830   6.06  15.17  30.87  13.74   7.71  21.57  17.67  38.27  10.66  11.89   
803  26.73  17.72   2.00  18.87  25.07  11.10  15.30   6.00  22.12  20.73   

       831    808    802    805    825    806    830    803    829  
829   3.36  16.92  33.72  24.82   7.46  23.22   2.38  30.52   0.00  
827   7.04  13.13  29.93  21.03   3.65  19.43   6.06  26.73   4.14  
821  16.15   4.12  20.92  12.02   6.22  10.42  15.17  17.72  13.26  
804  31.84  12.30   5.20   4.40  21.68   6.10  30.87   2.00  28.92  
822  14.68   5.27  22.07  13.17   4.76  11.57  13.74  18.87  11.81  
826   8.69  11.47  28.27  19.37   2.20  17.77   7.71  25.07   5.81  
807  22.54   2.90  14.30   5.40  12.38   3.80  21.57  11.10  19.62  
809  18.64   1.70  18.50   9.60   8.48   8.00  17.67  15.30  15.72  
801  39.24  19.60   2.80  11.70  29.08  13.40  38.27   6.00  36.32  
824  11.64   8.52  25.32  16.42   1.71   7.80  10.66  22.12   8.77  
823  12.87   7.13  23.93  15.03   2.94  13.43  11.89  20.73  10.00  
831   0.00  19.84  36.64  27.74  10.33  26.14   1.39  33.44   3.36  
808  19.84   0.00  17.10   8.20   9.68   6.60  18.87  13.90  16.92  
802  36.64  17.10   0.00   9.20  26.48  10.90  35.67   3.40  33.72  
805  27.74   8.20   9.20   0.00  17.58   2.00  26.77   6.00  24.82  
825  10.33   9.68  26.48  17.58   0.00  15.98   9.35  23.28   7.46  
806  26.14   6.60  10.90   2.00  15.98   0.00  25.17   7.70  23.22  
830   1.39  18.87  35.67  26.77   0.00  25.17   0.00  32.47   2.38  
803  33.44  13.90   3.40   6.00  23.28   7.70  32.47   0.00  30.52